File: mobileproject-backend/src/main/java/com/mobileproject/mobileprojectbackend/transaction/Transaction.java
package com.mobileproject.mobileprojectbackend.transaction;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.Instant;

/**
 * Entity đại diện cho một giao dịch tài chính (thu/chi) của cặp đôi.
 *
 * <p>Lưu trong MongoDB collection {@code transactions}.
 * Mỗi giao dịch gắn với một cặp đôi thông qua {@code coupleId}.</p>
 *
 * <h3>Indexes:</h3>
 * <ul>
 *   <li>{@code id_couple} – indexed, truy vấn nhanh theo cặp đôi</li>
 * </ul>
 *
 * <h3>Mối quan hệ:</h3>
 * <ul>
 *   <li>{@code coupleId} → tham chiếu đến {@code CoupleInfo.id}</li>
 * </ul>
 */
@Document(collection = "transactions")
public class Transaction {

    /**
     * ID duy nhất của giao dịch (MongoDB ObjectId).
     */
    @Id
    private String id;

    /**
     * ID của cặp đôi sở hữu giao dịch này.
     * Được index để truy vấn nhanh.
     */
    @Indexed
    @Field("id_couple")
    private String coupleId;

    /** Số tiền giao dịch (đơn vị: VND). */
    private Long amount;

    /** Loại giao dịch: {@link TransactionType#INCOME} hoặc {@link TransactionType#EXPENSE}. */
    private TransactionType type;

    /** Danh mục giao dịch (ví dụ: "Ăn uống", "Di chuyển", "INCOME"). */
    private String category;

    /** Ghi chú thêm cho giao dịch. */
    private String note;

    /** Thời điểm tạo giao dịch (UTC). */
    @Field("created_at")
    private Instant createdAt;

    /**
     * Constructor mặc định bắt buộc bởi MongoDB driver.
     */
    public Transaction() {
    }

    /**
     * Tạo giao dịch mới với thời gian tạo là thời điểm hiện tại.
     *
     * @param coupleId ID cặp đôi
     * @param amount   số tiền (VND)
     * @param type     loại giao dịch (INCOME / EXPENSE)
     * @param category danh mục
     * @param note     ghi chú
     */
    public Transaction(String coupleId, Long amount, TransactionType type, String category, String note) {
        this.coupleId = coupleId;
        this.amount = amount;
        this.type = type;
        this.category = category;
        this.note = note;
        this.createdAt = Instant.now();
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getCoupleId() {
        return coupleId;
    }

    public void setCoupleId(String coupleId) {
        this.coupleId = coupleId;
    }

    public Long getAmount() {
        return amount;
    }

    public void setAmount(Long amount) {
        this.amount = amount;
    }

    public TransactionType getType() {
        return type;
    }

    public void setType(TransactionType type) {
        this.type = type;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getNote() {
        return note;
    }

    public void setNote(String note) {
        this.note = note;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }
}

File: mobileproject-android/app/src/main/java/com/example/mobileproject/domain/entity/Transaction.kt
package com.example.mobileproject.domain.entity

/**
 * Domain entity đại diện cho một giao dịch tài chính trong ví chung của cặp đôi.
 * Mỗi giao dịch ghi lại một lần thu hoặc chi, phục vụ quản lý tài chính và phân tích.
 *
 * @property id Định danh duy nhất của giao dịch
 * @property coupleId Định danh cặp đôi sở hữu giao dịch này
 * @property amount Số tiền giao dịch (đơn vị: VNĐ), luôn dương
 * @property type Loại giao dịch: INCOME (thu) hoặc EXPENSE (chi)
 * @property category Danh mục chi tiêu/thu nhập (ví dụ: "Ăn uống", "Lương")
 * @property note Ghi chú mô tả giao dịch do người dùng nhập
 * @property createdAt Thời điểm tạo giao dịch dạng chuỗi ISO
 */
data class Transaction(
    val id: String,
    val coupleId: String,
    val amount: Long,
    val type: TransactionType,
    val category: String,
    val note: String,
    val createdAt: String
)

/**
 * Enum phân loại loại giao dịch.
 *
 * - INCOME: Thu nhập — tiền được thêm vào ví (lương, thưởng, hoàn tiền)
 * - EXPENSE: Chi tiêu — tiền được trừ khỏi ví (mua sắm, ăn uống, giải trí)
 */
enum class TransactionType {
    INCOME,
    EXPENSE
}

/**
 * Domain entity chứa kết quả trả về từ API khi tạo hoặc truy vấn giao dịch.
 * Các trường nullable phản ánh việc server có thể không trả về đầy đủ thông tin
 * trong mọi trường hợp (ví dụ: khi tạo thành công chỉ trả về id và số dư mới).
 *
 * @property success True nếu thao tác thành công
 * @property message Thông điệp phản hồi từ server
 * @property transactionId Định danh giao dịch được tạo, null nếu thao tác thất bại
 * @property amount Số tiền của giao dịch
 * @property type Loại giao dịch dưới dạng chuỗi (cần parse thành [TransactionType])
 * @property category Danh mục giao dịch
 * @property note Ghi chú giao dịch
 * @property totalBalance Số dư ví sau giao dịch, dùng để cập nhật UI real-time
 * @property createdAt Thời điểm tạo giao dịch
 */
data class TransactionResponse(
    val success: Boolean,
    val message: String,
    val transactionId: String?,
    val amount: Long?,
    val type: String?,
    val category: String?,
    val note: String?,
    val totalBalance: Long?,
    val createdAt: String?
)
